### ISSUE TITLE
States of tabs restored incorrectly

### ISSUE THREAD

**Issue by anandmahesh** [michael-rapp/ChromeLikeTabSwitcher#6]
I am developing Browser based on this library . Currently i am facing issue regarding reinflation of tabs on minimise and reopen of App. I am using WebView inside Tab and this reinflation cause the webview state loss . I already use webview store and restore state with OnSavedInstance of Decorator but in some cases i required webview to be inflated in Tab and this Tab must not be reinflated on minimise and reopen or app switch .

I observed some logs in Android Studio regarding removall of tabview from recyclerview like this code in onDetachLayout method of class PhoneTabSwitcherLayout : 

childViewRecycler.removeAll();
childViewRecycler.clearCache();
viewRecycler.removeAll();
viewRecycler.clearCache();
recyclerAdapter.clearCachedPreviews();

If I am taking it right removall and cache clear tend to reinflation of Tabs and calling of this method onInflateLayout in same class for reintialize of Objects like childViewRecycler ,viewRecycler and RecyclerAdapter etc.

I need solution for this state loss .
Kindly help.

**Comment by michael-rapp** [comment]
You are right with the assumption, that the method calls you posted cause the tabs to be re-inflated. This is intended, not only on orientation changes, but also when minimizing and reopening the app. This is not the default behavior for views in Android. Usually, their state is retained when minimizing. However, I ran into some serious issues with this behavior and therefore decided to always re-inflate the tabs. I agree, that this is not the optimal solution, but it is how it's currently supposed to be.

If I understand you correctly, the web view inside of your tabs is not always restored correctly. Maybe this is a bug, which prevents the Decorator's onRestoreInstanceState-method from being called correctly, or it is called with a wrong or corrupted Bundle as a parameter.

Unfortunately, I am currently very busy and don't have the time to investigate on this. However, if you want to do some debugging on your own, I am willing to help whenever I am able to. I would also appreciate to receive some patches fixing issues like that.

**Comment by michael-rapp** [comment]
I just did some testing using version 0.1.3 of the library and the corresponding example app. There is definitely a bug in how the states of tabs are restored. It seems like some tabs are restored using the states of other tabs. Is this in accordance with your observations?

**Comment by anandmahesh** [comment]
"It seems like some tabs are restored using the states of other tabs " ...Yes I also observed this behaviour when you use same viewType for particular type Tab like Web View Tab ( if this webview are loaded with webpages ) then adding multiple tab of this ViewType in tabswitcher and switching between this same type Tab trigger this behaviour . Thats why i suggest counter in this issue for ViewType : "https://github.com/michael-rapp/ChromeLikeTabSwitcher/issues/5" .

WVT - Web View Tab

For Example from above bheaviour if WVT1 is loaded with url1 and WVT2 is loaded with url2 so on...their are 10 WVTs then switching tabs in tabswitcher inflate the wrong WVT at wrong index ..like WVT2 inflate WVT1 webview loaded with url1 ,so it is same as restore using the states of other tabs.

Above is the Web View Tab Problem I face .

**Comment by anandmahesh** [comment]
"If I understand you correctly, the web view inside of your tabs is not always restored correctly. Maybe this is a bug, which prevents the Decorator's onRestoreInstanceState-method from being called correctly, or it is called with a wrong or corrupted Bundle as a parameter."

For handling WebView State i have separate Logic Implemented beacuse of same bheaviour of reinflation of tabs at different instance . I am not able to grab calling conditions of savedinstance state its so frequently.
I only using OnSavedInstanceState for getting the webview state inside Bundle and with the help of separate logic in onShowTab i restore the webview state after webview is intialize and i also have some my own conditions to handle restore is required or not .

**Comment by anandmahesh** [comment]
"Unfortunately, I am currently very busy and don't have the time to investigate on this. However, if you want to do some debugging on your own, I am willing to help whenever I am able to. I would also appreciate to receive some patches fixing issues like that."

**Comment by michael-rapp** [comment]
> "It seems like some tabs are restored using the states of other tabs " ...Yes I also observed this behaviour 

It think this is the first thing we should track down. It has nothing to do with web views and is easily reproducable using the library's example. Just insert some text into the EditText widgets, which are contained by the tabs.

**Comment by anandmahesh** [comment]
"Unfortunately, I am currently very busy and don't have the time to investigate on this. However, if you want to do some debugging on your own, I am willing to help whenever I am able to. I would also appreciate to receive some patches fixing issues like that."

Ok i need to understand the code in depth , if possible can you please provide some Flow Chart or Documentation for fast understanding of code in library.

**Comment by michael-rapp** [comment]
There is no documenation such as a flow chart for the library. And I don't have any detail in mind to create such a chart in few time. I understand, that the code is really complex and that its hard to understand. Debugging through the relevant parts of it would be the best approach I think. I also have to do it, to understand what is going on here.

Dont't mind, if you are unable to understand the code in detail. I am planning to work on this issue during the next days. It is really a major issue and I want it to be resolved.

**Comment by michael-rapp** [comment]
I just took a fast look at the issue and debugged it using the example app. It turned out, that it isn't really a bug in the library, but in the example app itself. I will explain it in the following. Maybe it also applies to your app.

Some of the example app's tabs contain `EditText` widgets. If such a tab is removed, the `EditText` is kept in memory to be reused later. If a text has been entered, it is still stored in the `EditText`. When a new tab is added, the `EditText` is eventually reused with the text still present. 

The removal and addition of tabs happens in background when scrolling through the tabs. This may cause the contents of tabs to appear at different places where they don't belong to. The actual issue is, that the example app's decorator implementation does not reset the text of reused views. By adding the following lines of code to the decorator's `onShowTab`-method, the issue can be fixed:

```java
public void onShowTab(@NonNull final Context context,
                              @NonNull final TabSwitcher tabSwitcher, @NonNull final View view,
                              @NonNull final Tab tab, final int index, final int viewType,
                              @Nullable final Bundle savedInstanceState) {
    // Adapt the view to the appearance of the tab...
          
    EditText editText = findViewById(android.R.id.edit); // This is the EditText. It might be reused
    
    // The following if needs to be added
    if (savedInstanceState == null) {
        // There is now saved state for the tab, i.e. it is show for the first time and its text should be empty
        editText.setText(null);
    } else {
        // It is not necessary to do anything in this case, because the view state is implicitly restored by the library, if a saved state is present
    }
}
```

**Comment by michael-rapp** [comment]
@anandmahesh Please tell me, if the solution above solves your problem.

**Comment by michael-rapp** [comment]
I just released version 0.1.4 which fixes said issue in the example app.

**Comment by anandmahesh** [comment]
Its not the solution .

**Comment by anandmahesh** [comment]
Its not the Solution as i think .

- Initial State of Tab : when tab just added in the Tab-Switcher .
- Current State of Tab : when tab is added and some working is done on Tab like as your example some text is type in Edit Text .

Bug Points in library are:

According to as you said in previous Issue "When a tab is removed, its view is not deleted, but kept in memory to be reused when adding a tab with the same view type." 

1. if Tab is added of a particular viewtype (say 1) in library then removal of this tab saved the current state of Tab for viewtype 1. If after some time user add the viewtype 1 tab in tabswitcher it restore the old viewtype 1 tab current state its not right according to my view.
If its possible please try to save the viewtype 1 Initial State as the View in cache for fast performance .

"The removal and addition of tabs happens in background when scrolling through the tabs. This may cause the contents of tabs to appear at different places where they don't belong to."

2. if its possible can scrolling be bind with Tab-Index then may be inflation of view is on right position and Initial State must be taking into consideration for this background behaviour.

**Comment by michael-rapp** [comment]
> if Tab is added of a particular viewtype (say 1) in library then removal of this tab saved the current state of Tab for viewtype 1. If after some time user add the viewtype 1 tab in tabswitcher it restore the old viewtype 1 tab current state its not right according to my view.
If its possible please try to save the viewtype 1 Initial State as the View in cache for fast performance .

I have to check, if this is actually the current behavior. If it is, this is a problem indeed. However, I don't like the idea of storing the initial states of tabs as this takes additional memory. I would leave it in the responsibility of the decorator implementation to reset tabs. For this reason I am thinking about providing three methods, which must be overridden in a decorator:
- One to inflate view (already exists)
- One to adapt a view to a particular tab (already exists)
- One to reset a view to the initial state (called when reusing a view)

> if its possible can scrolling be bind with Tab-Index then may be inflation of view is on right position and Initial State must be taking into consideration for this background behaviour.

I'm not sure, if I understand this correctly. What you propose is to remove a view when the corresponding tab is scrolled out of the visible area and keep it in memory without reusing it for other tabs? So the same view will be used when the tab becomes visible again? Of course this would make restoring the tab's state obsolete when scrolling. But it also requires to inflate n views, if n tabs are present, right?

**Comment by anandmahesh** [comment]
"I'm not sure, if I understand this correctly. What you propose is to remove a view when the corresponding tab is scrolled out of the visible area and keep it in memory without reusing it for other tabs? So the same view will be used when the tab becomes visible again? Of course this would make restoring the tab's state obsolete when scrolling. But it also requires to inflate n views, if n tabs are present, right?"

No , leave my opinion in scroll bheaviour , i only observe that bheaviour small time then for developing Browser i find workaround mention below .

the problem we are facing on scroll can be rsolved if you always use different viewtype for creating the tab it also includes inflation of same xml for particluar range of viewtype as i required , so thats how i overcome the scroll issue.

for further assistance i am going to look into this scroll bheaviour in deep and update you as soon as possible .

currently i am focusing on solution for minimize or switch app reinflation.

if you can mention which issue require priority or some dependency they both have i can switch to that issue first..?

**Comment by michael-rapp** [comment]
@anandmahesh Did you made any progress on this issue? I just dived into the tab-restoring mechanism of the library once again and it works well for me. I could not find any issues except for https://github.com/michael-rapp/ChromeLikeTabSwitcher/issues/7, which I will try to fix.

**Comment by yuchuangu85** [comment]
I find this problem when we use your project, because the project recycle to use the view, so if one tab is changed ,the recycle view will be changed. I solve it by set the parameter “getViewTypeCount” return 10, because the max count of the tabs is 10. sorry, I am Chinese, my English is poor。

**Comment by michael-rapp** [comment]
Setting the maximum view count to the maximum number of tabs is a really bad workaround. It affects the performace in a negative way.

As I said earlier, I double-checked the implementation and wasn't able to find a bug. As the recycling mechanism works fine in the example app, your implementation is likely to be buggy. You should pay attention to the following important points:

* If no `savedInstanceState` is provided to the decorator's `onShowTab`-method, all views should be reset to the default state (see previous comment in https://github.com/michael-rapp/ChromeLikeTabSwitcher/issues/6#issuecomment-333592651).
* Previously saved states are not removed from memory (this enables to restore the state of previously removed tabs when they are re-added later). If you only encounter problems when adding new tabs (not when scrolling), the reason might be, that the hash code of a new `Tab` is equal to the hash code of a previously removed tab. Although this is unlikely to happen, it might be the case if the newer `Tab` object is located at the same memory address as the original one. If that's the case you should extend the class `Tab` and override the `hashCode` and `equals` method in order to ensure that such things cannot happen).

Maybe it helps, if you imagine the tab switcher to be a `ListView`. The idea behind its view recycling mechanism is identical.

**Comment by yuchuangu85** [comment]
Thank you very much！I'll check it again.

**Comment by michael-rapp** [comment]
I just released version 0.1.6, which prevents the states of previously removed tabs from being reused when adding a new tab that happens to have the same hash code. This makes the second point in https://github.com/michael-rapp/ChromeLikeTabSwitcher/issues/6#issuecomment-353510051 obsolete.

**Comment by michael-rapp** [comment]
I will now close this issue. If someone encounters related problems, please create a new bug report with detailed information.
